dspace-swordv2/src/main/java/org/dspace/sword2/VersionManager.java

```diff
@@ -10,75 +10,95 @@
 import org.dspace.authorize.AuthorizeException;
 import org.dspace.content.Bitstream;
 import org.dspace.content.Bundle;
+import org.dspace.content.BundleBitstream;
 import org.dspace.content.Item;
+import org.dspace.content.factory.ContentServiceFactory;
+import org.dspace.content.service.BitstreamService;
+import org.dspace.content.service.BundleService;
+import org.dspace.content.service.ItemService;
 import org.dspace.core.ConfigurationManager;
+import org.dspace.core.Context;
 
 import java.io.IOException;
 import java.sql.SQLException;
 import java.text.SimpleDateFormat;
 import java.util.Date;
+import java.util.Iterator;
+import java.util.List;
 
 public class VersionManager
 {
-	public void removeBundle(Item item, String name)
+	protected ItemService itemService = ContentServiceFactory.getInstance().getItemService();
+	protected BundleService bundleService = ContentServiceFactory.getInstance().getBundleService();
+	protected BitstreamService bitstreamService = ContentServiceFactory.getInstance().getBitstreamService();
+
+	public void removeBundle(Context context, Item item, String name)
 			throws SQLException, AuthorizeException, IOException
 	{
 		boolean keep = ConfigurationManager.getBooleanProperty("swordv2-server", "versions.keep");
-		Bundle[] bundles = item.getBundles(name);
+		List<Bundle> bundles = item.getBundles();
```

Comment:
@aschweer It will also cause an issue here, I didn't yet notice this one apparently. The problem is that in the call below: "this.removeBundle(context, item, b, keep)" actually calls an "itemService.removeBundle(context, item, source)" method.
Which will result in an alteration of the "bundles" list which are now looping over. This will lead to the ConcurrentModificationException.

The problem isn't related to a service call or not, it occurs because hibernate returns persisted lists and deleting an object results in a list alteration.
We used to work with just value based arrays which didn't have this problem since the object would still exist, when we delete an object using persisted lists the object is destroyed when it is deleted (so it is really gone). Since the object is gone the list loop will crash since the list picks up on this change.

Hope this helps.